### PR TITLE
Better grl

### PR DESCRIPTION
**Pull request by thejevans**

### CODE REVIEW

File: mla/analysis.py [hunk 1]
```diff
@@ -147,11 +145,19 @@ def produce_trial(analysis: Analysis, flux_norm: float = 0,
...
+        background['time'],
+        analysis.model.background_time_profile,
+    )
 
-    if flux_norm > 0 or n_signal_observed > 0:
+    if flux_norm > 0 or n_signal_observed is not None:
         signal = analysis.model.inject_signal_events(analysis.source,
                                                      flux_norm,
                                                      n_signal_observed)
+        signal['time'] = analysis.model.scramble_times(
```
Comment (jasonfan1997) [inline mla/analysis.py:157]: Why would injected signals require grl to weight time distribution?

**Comment by jasonfan1997** [comment]
I think the idea of distributing events base on the grl rate is good, but if we don't change the corresponding time likelihood maybe it could cause a false positive. Also, I think the injected signal events should not weight based on the grl rate 
The original idea of disabling the grl filter is to stop filtering out injected signal events when the n_signal_observed is given so it should still be keep.

File: mla/_models.py [hunk 2]
```diff
@@ -455,22 +458,39 @@ def inject_signal_events(
 
         return signal
 
-    def grl_filter(self, events: np.ndarray) -> np.ndarray:
+    def _init_grl_rates(self, grl: np.ndarray, data: np.ndarray) -> np.array:
```
[the remark this answers is not in the document]
Comment (thejevans) [reply mla/_models.py:461]: you're right. will update

File: mla/analysis.py (hunk 1 above)
Comment (thejevans) [reply mla/analysis.py:157]: you're right. I should have a grl_weight flag or something to turn this on for background only

Comment (jasonfan1997) [reply mla/analysis.py:157]: I think we would still need to remove signal events that are outside grl after assign time based on the signal time distribution(and an option to disable the filter)

File: mla/_test_statistics.py [hunk 3]
```diff
@@ -183,7 +187,7 @@ def i3_ts(sob: np.ndarray, prepro: Preprocessing,
     ) + prepro.n_dropped * np.log(1 - ns_ratio))
 
 
-def cal_ns_ratio(sob: np.array, n_dropped: int, iterations: int = 30) -> float:
+def cal_ns_ratio(sob: np.array, n_dropped: int, iterations: int = 10) -> float:
```
[the remark this answers is not in the document]
Comment (thejevans) [reply mla/_test_statistics.py]: fixed

File: mla/_models.py [hunk 4]
```diff
@@ -571,7 +571,7 @@ def inject_background_events(self) -> np.ndarray:
         n_background_observed = np.random.poisson(self._n_background)
         background = np.random.choice(self._data, n_background_observed).copy()
         background['ra'] = np.random.uniform(0, 2 * np.pi, len(background))
-        background['time'] = self._background_time_profile.random(
+        background['time'] = self.background_time_profile.random(
```
Comment (jasonfan1997) [inline mla/_models.py:574]: Why the scramble time part is not in here , replacing this line? You are doing time sampling twice here.

File: mla/_models.py [hunk 5]
```diff
@@ -617,6 +617,6 @@ def inject_signal_events(
                 signal['trueRa'], signal['trueDec'])
 
         # Randomize the signal time
-        signal['time'] = self._signal_time_profile.random(size=len(signal))
+        signal['time'] = self.signal_time_profile.random(size=len(signal))
```
Comment (jasonfan1997) [inline mla/_models.py:620]: Same as above.

Comment (thejevans) [reply mla/_models.py]: fixed.

File: mla/_models.py (hunk 4 above)
Comment (thejevans) [reply mla/_models.py]: fixed

File: mla/test_statistics.py [hunk 6]
```diff
@@ -51,12 +61,31 @@ def _preprocess(self, event_model: _models.EventModel,
...
+    if 'gamma' in params.dtype.names:
+        gamma = params['gamma']
+    else:
+        gamma = prepro.gamma
+    return np.exp([spline(gamma) for spline in prepro.splines])
+
+
+def i3_sob(params: np.ndarray, prepro: I3Preprocessing) -> np.array:
+    """Docstring"""
+    sob = prepro.sob_spatial
```
Comment (jasonfan1997) [inline mla/test_statistics.py:78]: Here need to change to prepro.sob_spatial.copy()

File: mla/threeml/test_statistics.py [hunk 7]
```diff
@@ -31,36 +30,10 @@ class ThreeMLPreprocessor(_test_statistics.TdPreprocessor):
...
-    sob = prepro.sob_spatial * \
-        _test_statistics.cal_sob_time(params, prepro) * sob_energy
-
-    ns_ratio = None
-    if 'ns' in temp_params.dtype.names:
-        ns_ratio = temp_params['ns'] / prepro.n_events
-    return _test_statistics.i3_ts(sob, prepro, return_ns, ns_ratio)
+def get_sob(params: np.ndarray, prepro: ThreeMLPreprocessing) -> np.array:
+    """Docstring"""
+    sob = prepro.sob_spatial
```
Comment (jasonfan1997) [inline mla/threeml/test_statistics.py:35]: Here also

File: mla/test_statistics.py (hunk 6 above)
Comment (thejevans) [reply mla/test_statistics.py]: thanks! fixed

File: mla/threeml/test_statistics.py (hunk 7 above)
Comment (thejevans) [reply mla/threeml/test_statistics.py]: fixed

**Pull request merged**